[Bug] Resume with multiple particle_types

cgeudeker opened the pull request:
**Describe the PR**
When resuming a simulation with multiple particle_types, a full set of particles would be generated for each particle type. Additionally, it seems that each instance of the same particle type would be loaded in full, though subsequent generation would likely replace the previous. As such the following modifications have been made:

* In `ParticlePODTypeName`, each distinct particle type has been assigned a specific POD type name, though these all lead to the eventual use of the "particles" POD type for file reading and writing.
* In `initialise_particle_types`, a modification has been made to only add distinct particle types.
* New .h5 files for each particle_type will be generated as a result, ensuring no multiple generations of the same particle upon resume.

**Related Issues/PRs**
N/A

**Additional context**
N/A

aquinorayme committed:
added rotation force input in mpm_base

aquinorayme committed:
Passed rotation parameters to compute_forces

aquinorayme committed:
Defined compute_forces to accept and apply rotation arguments

aquinorayme committed:
Implemented centrifugal and Coriolis force calculations

aquinorayme committed:
Updated scheme tests to include dummy rotation arguments

aquinorayme committed:
Applied clang-format fixes

aquinorayme committed:
Fixed formatting according to cppcheck

cgeudeker committed:
Modifications to save different particle types to different files. Ensures that resume does not load particle types multiple times. Also only adds unique particle types to particle_types_. Not clanged. Not tested.

cgeudeker committed:
Clang formatted. Passes mpm tests.

bodhinandach commented on include/solvers/mpm_base.tcc:
```diff
@@ -338,7 +338,9 @@ void mpm::MPMBase<Tdim>::initialise_particle_types() {
-    particle_types_.insert(particle_type);
+    // Insert only unique particle types
+    if (particle_types_.find(particle_type) == particle_types_.end())
+      particle_types_.insert(particle_type);
```
I thought std::set only stores unique entries, i.e., it does not store duplicates. If so, then this check is unnecessary.

cgeudeker replied:
Maybe you're right, documentation seems to indicate that. I thought when I was testing it I saw it go through P2D twice though. Will confirm and remove if unnecessary.

bodhinandach reviewed (commented):
@cgeudeker Thanks for finding and fixing this. Perhaps it is good to make some tests, if it is not too difficult?

cgeudeker commented:
> @cgeudeker Thanks for finding and fixing this. Perhaps it is good to make some tests, if it is not too difficult?

Yes will work on a couple.

cgeudeker committed:
New save structure for .h5 files. Not yet passing all tests.

cgeudeker committed:
New save structure for .h5 files. Not yet passing all tests.

cgeudeker committed:
Simplified write-read setup for HDF5. No updated test cases.

cgeudeker commented on include/mesh/mesh.h:
```diff
@@ -432,21 +432,14 @@ class Mesh {
-  //! \retval status Status of reading HDF5 output
-  bool read_particles_hdf5(const std::string& filename,
-                           const std::string& type_name,
-                           const std::string& particle_type);
```
Becomes the read_hdf5() function in mpm_base.

adrienmartinezZ committed:
Cleaned the format for parameter descriptions.
- Added short descriptions for input parameters in map_rotation_force in particle.h
- Spelled "stress" correctly in the description of initial_stress in particle.h
- Added description comment for rotation_forces_ in mpm_base.h
- Deleted extra "/" in compute forces description in mpm_scheme.h

aquinorayme committed:
Applied clang-format fixes

aquinorayme committed:
Fixed Clang formatting errors

aquinorayme committed:
Small test for Rotation Forces v.1

aquinorayme committed:
Applied clang format to v.1

aquinorayme committed:
Added 3D version of json error

aquinorayme committed:
Small test v.2

cgeudeker committed:
Add test case for starting and resuming an implicit simulation with multiple particle types

aquinorayme committed:
Changed map_rotation_forces to rotation_function

aquinorayme committed:
Removed comments

aquinorayme committed:
Updated rotation force tests with boolean pass/fail

aquinorayme committed:
Applied clang format

aquinorayme committed:
Initialize mesh and materials before applying loads in rotation tests

bodhinandach commented on include/mesh/mesh.tcc:
```diff
@@ -1569,21 +1569,28 @@ std::vector<std::array<mpm::Index, 2>> mpm::Mesh<Tdim>::particles_cells()
-    auto pod = std::static_pointer_cast<mpm::PODParticle>((*pitr)->pod());
-    particle_data.emplace_back(*pod);
+    if (particle_type == (*pitr)->type()) {
+      auto pod = std::static_pointer_cast<mpm::PODParticle>((*pitr)->pod());
```
This casting to `mpm::PODParticle` means that the different particle types should have the same POD t,ype right? That means single-phase particles cannot be output together with two-phase particles since they have different POD types. Is that the intention? I guess, we are not going to run a scenario of both multi-phase particles and single-phase particles simulated at the same time, right?

bodhinandach commented on include/solvers/mpm_base.tcc:
```diff
@@ -560,22 +550,47 @@ bool mpm::MPMBase<Tdim>::checkpoint_resume() {
+//! Read HDF5 files
+template <unsigned Tdim>
+void mpm::MPMBase<Tdim>::read_hdf5() {
+  // Read hdf5 file for single phase particle
```
Comment may be misleading

bodhinandach commented on include/solvers/mpm_base.tcc:
```diff
@@ -526,17 +526,7 @@ bool mpm::MPMBase<Tdim>::checkpoint_resume() {
-      // Load particle information from file
-      mesh_->read_particles_hdf5(particles_file, attribute, ptype);
-    }
+    this->read_hdf5();
```
I am not too sure what the necessity is to bundle them into a single function. Is this just to improve readability? Or is there any specific efficiency reason?

bodhinandach reviewed (commented):
@cgeudeker Thanks for the effort. Overall, looks good to me. But I have some questions that I would like to understand.

aquinorayme committed:
Renamed function

aquinorayme committed:
First est in particle_test.cc (v.1)

aquinorayme committed:
Applied clang format for changes

aquinorayme committed:
Added explicit physics validation for centrifugal and coriolis forces

aquinorayme committed:
Flipped signn of expected values

aquinorayme committed:
Attempt v.2

aquinorayme committed:
Removed  comments

bodhinandach commented on include/mesh/mesh.tcc:
```diff
@@ -1593,29 +1600,32 @@ bool mpm::Mesh<Tdim>::write_particles_hdf5(const std::string& filename) {
-    particle_data.emplace_back(*pod);
+    if (particle_type == (*pitr)->type()) {
+      auto pod =
+          std::static_pointer_cast<mpm::PODParticleTwoPhase>((*pitr)->pod());
```
Same here....

aquinorayme committed:
Removed irrelevant comments

aquinorayme committed:
Replaced std::runtime error

aquinorayme committed:
Removed this-> as discussed, added "using", and applied clang format

aquinorayme committed:
Fixed mpm_explicit.h removed line

aquinorayme committed:
Applied clang format (v.3)

aquinorayme committed:
Changed to warn instead of error

adrienmartinezZ committed:
Added gravity and rotational forces ramping
-Added ramping_time parameter to mpm_base.h
-Implemented ramping logic in mpm_base.tcc "initiatlise_loads()" with negative throws + same ramping throws
-Apply ramping multipliers to gravity and rotation forces in mpm_explicit.tcc "update_particle_velocities()"
-Initialized write_json_ramping() function to write_mesh_particles.h and write_mesh_particles.cc for test
-Add ramping test cases to mpm_explicit_usf_test.cc

Test Cases for Ramping:
- Case 1: Ramping gravity with t > 0s
- Case 2: Ramping gravity with t < 0s (error case 1)
- Case 3: Ramping rotation with t > 0s
- Case 4: Ramping rotation with t < 0s (error case 2)
- Case 5: Ramping both gravity and rotation at same time (error case 3)
- Case 6: No ramping (error case 4)

adrienmartinezZ committed:
Added 3D case to test ramping
-3D case to mpm_explicit_usf_test.cc for gravity ramping

adrienmartinezZ committed:
Changing 3D case to solve() to run.

adrienmartinezZ committed:
Added tests to cover code and more warnings to check.

adrienmartinezZ committed:
Reverting warnings tests to cover branches for 2D and 3D cases.

bodhinandach committed:
Add json search

bodhinandach committed:
modifications including two different ramping times for gravity and rotation force

bodhinandach committed:
pushing config.yml to check clang-format version

bodhinandach committed:
Change yml to see if we can update clang-format version

bodhinandach committed:
Add change direction for user instead of global

bodhinandach committed:
Replaced 6 deprecated options with their modern clang-format 18 equivalents

adrienmartinezZ committed:
Independent multipliers for gravity and rotational ramping
Testing for 2D and 3D cases with the following combinations:
-Gravity only
-Rotational ramping only
-Both gravity and rotational ramping

bodhinandach committed:
All clang-format

bodhinandach committed:
Change version in yml

bodhinandach committed:
Revert clang-format exec calling

bodhinandach committed:
Another attempt to fix yml

adrienmartinezZ committed:
Fixed formatting for clang-tidy issues in mpm_base.tcc

aquinorayme committed:
Added tests to cover IO missing file error and time-ramping multiplier cutting off

aquinorayme committed:
Updated explicit solver Catch2 tests to execute a full solve() cycle instead of only initialise_loads()

aquinorayme committed:
Refactored ramping JSON generation to execute full solve loops

aquinorayme committed:
Refactored ramping JSON generation to fix exception coverage

aquinorayme committed:
Refactored ramping JSON generation to fix exception coverage

aquinorayme committed:
Added 3D clamping tests and executed full solve loops

aquinorayme committed:
Fixed final coverage gaps for base warnings, IO search, and Eigen operations

aquinorayme committed:
Combined redundant ramping and clamping test cases to reduce lines of code

bodhinandach committed:
Modify author contribution names

cgeudeker committed:
Updating code comments for accuracy and clarity.

cgeudeker committed:
Updated code comments with clang format.

cgeudeker committed:
Final part of rebase?

cgeudeker committed:
Mesh.h rebase

cgeudeker committed:
Fixed for main branch updates

cgeudeker committed:
Merge remote-tracking branch 'origin/master' into bugfix/resumeParticleType

cgeudeker committed:
Retry merge main branch.

cgeudeker commented:
@bodhinandach cppcheck looks like its indicating a clang format error but it passes the clang test? Not sure what the difference is or what to fix to move forward. Note I had run format.sh prior to pushing the modifications.

cgeudeker committed:
Correct clang format version

cgeudeker committed:
Manual clang fix.

cgeudeker committed:
Manual Clang fix part 2.

bodhinandach reviewed (approved):
Thanks @cgeudeker , I am good with this.
